doc: design document for rbd mirroring (#1621)
ceph/ceph-csi: docs/design/proposals/rbd-mirror.md

    imageFeatures: layering
    imageFormat: "2"
    imageName: csi-vol-0c23de1c-18fb-11eb-a903-0242ac110005
    journalPool: replicapool
    pool: replicapool
    radosNamespace: ""
  volumeHandle: 0001-0009-rook-ceph-0000000000000002-0c23de1c-18fb-11eb-a903-0242ac110005
```

> **VolumeHandle** is the unique volume name returned by the CSI volume plugin’s
CreateVolume to refer to the volume on all subsequent calls.

Once the static PVC is created on the secondary cluster, the Kubernetes User
can try delete the PVC,expand the PVC or mount the PVC. Incase of mounting
(NodeStageVolume) we will get the volume context in RPC call but not in the
Delete/Expand Request. In Delete/Expand RPC request only the VolumeHandle
(`clusterID-poolID-volumeuniqueID`) will be sent where it contains the encoded
information related to ClusterID and PoolID. The VolumeHandle is not useful in
secondary cluster as the PoolID and ClusterID always may not be the same.

> In this design document we will talk about new controller(rbdplugin
> controller) not replication controller, in next releases we will design the
> replication controller to perform mirroring operations. The rbdplugin

> > controller will runs as a sidecar in RBD provisisoner pod.

Reviewer: ```suggestion
> controller will run as a sidecar in RBD provisioner pod.
```

To solve this problem, We will have a new controller(rbdplugin controller)

> running as part of provisioner pod which watchers for the PV objects. When a PV

Reviewer: ```suggestion
running as part of provisioner pod which watches for the PV objects. When a PV
```

is created it will extract the required information from the PV spec and it
will regenerate the OMAP data and also it will generate a new VolumeHandle

> (`newclusterID-newpoolID-volumeuniqueID`) and it creates an OMAP object for
> mapping between old VolumeHandle and new VolumeHandle. whenever Ceph-CSI gets a
> RPC request with older VolumeHandle it will check is there any new VolumeHandle
> exists for the old VolumeHandle. If yes, it uses the new VolumeHandle for

Reviewer: ```suggestion
(`newclusterID-newpoolID-volumeuniqueID`) and it creates an OMAP object for
mapping between old VolumeHandle and new VolumeHandle. Whenever Ceph-CSI gets a
RPC request with older VolumeHandle, it will check if any new VolumeHandle
exists for the old VolumeHandle. If yes, it uses the new VolumeHandle for
```

internal operations (to get pool name, Ceph monitor details from the ClusterID
etc).

Currently, We are making use of watchers in node stage request to make sure
ReadWriteOnce (RWO) PVC is mounted on a single node at a given point in time.
We need to change the watchers logic in the node stage request as when we
enable the RBD mirroring on an image, a watcher will be added on a RBD image by
the rbd mirroring daemon.

> To solve the ClusterID problem, If the ClusterID is different on the second
> cluster the admin as to create a new ConfigMap for the mapped ClusterID's.

Reviewer: ```suggestion
To solve the ClusterID problem, If the ClusterID is different on the second
cluster, the admin has to create a new ConfigMap for the mapped ClusterID's.
```

> whenever Ceph-CSI gets a request it will check the ClusterID mapping exists and
> uses mapped ClusterID to get the information like Ceph monitors etc.

Reviewer: ```suggestion
Whenever Ceph-CSI gets a request, it will check if the ClusterID mapping exists and
uses the mapped ClusterID to get the information like Ceph monitors etc.
```

# RBD MIRRORING

RBD mirroring is a process of replication of RBD images between two or more
Ceph clusters. Mirroring ensures point-in-time, crash-consistent RBD images
between clusters, RBD mirroring is mainly used for disaster recovery (i.e.
having a secondary site as a failover). See Upstream [Ceph
documentation](https://docs.ceph.com/en/latest/rbd/rbd-mirroring) on RBD
mirroring for complete information.

## Architecture

> ![mirror](rbd-mirror.png)

Author: I don't have SVG now. I can add it later

## Design

Currently, ceph-csi generates its unique ID for each RBD image and stores the
corresponding PV name and the unique ID mapping, It creates the RBD image with
the unique ID and returns the encoded value which contains all the required
information for other operations. For mirroring, the same RBD image will be
mirrored to the secondary cluster. As the journal(omap data) is not mirrored to
the secondary cluster, The cephcsi cannot identify the RBD image corresponding
to the PV wihtout omap data.

> **Note:** It's expected that the user/admin will create the static PVC's on the

Author: yes correct

secondary site during the failover.

**Note:** when the static PVC created on the secondary site we cannot use the
volumeHandle to identify the omap data or the image anymore because as we have

> only PoolID and clusterID in the volumeHandle. we cannot identify the correct

Reviewer 2: consider formatting `PoolID` (with **C**apital?), `clusterID` and `volumeHandle` as these are names in the implementation

pool name from the poolID because pool name will remain the same on both clusters
but not the poolID even the clusterID can be different on the secondary

> cluster.

Author: Not always true, the pool name will be same not the clusterID

> **VolumeHandle** is the unique volume name returned by the CSI volume plugin’s

> CreateVolume to refer to the volume on all subsequent calls.

Reviewer 2: maybe *CreateVolume RPC response*

Author: This is taken from the kubectl pv.spec.volumeHandle output will leave as it is for now

Once the static PVC is created on the secondary cluster, the user can try

> delete the PVC,expand the PVC or mount the PVC. Incase of mounting (node stage)

Author: Yes the rbd image will be demoted on the second cluster or the admin as the force promote the rbd image on the secondary clster

> we will get the volume context in GRPC call but not in the Delete/Expand

Reviewer 2: Do not call them GRPC calls, that is a CSI internal implementation detail. Try to stick to the naming of operations or *CSI RPCs* as the spec does. (GRPC is just the format, *CSI RPC*, or *RPC call* is better in this doc)

Request. In Delete/Expand GRPC request only the
volumeHandle (`clusterID-poolID-volumeuniqueID`) will be sent where it contains
the encoded information related to clusterID and poolID. The volumeHandle is
not useful in secondary cluster as the poolID and clusterID always may not be
the same.

> To solve this problem, We will have a new controller running as part of

Reviewer 2: *We* -> *we*

provisioner pod which watchers for the PV objects. When a PV is created it will

> extract the required information from the PV spec and it will regenerate the

Author: > Few questions here:
> 
> *) Does this new controller run on both ( primary and secondary) clusters?
> 
yes it can run on both the clusters, functionality of the primary cluster wont be affected as the volumeHandler is already generated by CSI and omap data is already generated.

> *) Does this new controller run regardless of rbd mirroring is enabled in a setup ? iow, how are we controlling the deployment of this new controller from CSI deployment?
> 

It's up to the user to enable it or not, currently, it will be used only for mirroring use case but later the  user can make use of it later for expanding snapshot, and other operation which is not going to be covered here.

> *) What is the communication path between csi-rbdplugin container and the controller ( my assumption here is controller is just another container in the provisioner pod )? or it is going to be that, there is no communication between these, rather both works independently? if its independent, do we need to bundle it with provisioner pod?
> 
its nothing to do with the plugin as of now, it will run as a sidecar container in the provisioner pod.

> *) Is this new controller required, if CSI pod already know the `clusterID and Pool ID ` of the cluster ?
> 
It won't know anything, it will extract the information from the volumecontext

> *) It say `it will extract the required information` -> can we document what information is classified as `required` here?
> 
Sure I can update it not sure how much it will help, it will extract all the information present in the pv.spec.csi

> *) Controller expected to create a new VolumeHandle in format `newclusterID-newPoolID-volumeuniqueID` :
> 
> * Is it going to create a new volumehandle for all the `static PVs` in a cluster? ( regardless of whether the PV belong to mirrored image ?)
new volumehandle will not be used by the kubernetes, it will only be used by cephcsi internally for all operations.

> * Do we make use of certain properties of the PV to create this volumeUniqueID ? can you please list what are those properties if it tied ? or Is it  completely random UUID?
> 

It's the volumeHandle generated in the primary cluster.

> *) `whenever cephcsi gets a GRPC request with older volumeHandle it will check..` -> can you please confirm the check mentioned here is about the rbdplugin container?

yes rbdplugin will check the mapping between the volume handlers

omap data and also it will generate a new volumeHandle
(`newclusterID-newpoolID-volumeuniqueID`) and it creates an omap object for
mapping between old volumeHandle and new volumeHandle. whenever cephcsi gets a
GRPC request with older volumeHandle it will check is there any new
volumeHandle exists for the old volumeHandle. If yes, it uses the new
volumeHandle for internal operations (to get pool name, monitor details from
the clusterID etc).

> Currently, we are making use of watchers in node stage request to make sure RWO

Reviewer 2: expand *RWO*

PVC is mounted on a single node at a given point in time. We need to change the
watchers logic in the node stage request as when we enable the RBD mirroring on
an image, a watcher will be added on a RBD image by the rbd mirroring daemon.

To solve the clusterID problem, If the clusterID is different on the second

> cluster the admin as to create a new configmap for the mapped clusterID's.

Reviewer 2: *configmap* is a Kubernetes object, call it *ConfigMap*

whenever cephcsi gets a request it will check the clusterID mapping exists and

> uses mapped clusterID to get the information like monitors etc.

Reviewer 2: *monitors* -> maybe *Ceph monitors* or *Ceph MONs*?

> **This design does not cover the below items:**

Reviewer 2: Are there any other requirements to the environment, like firewalls, response times between Ceph clusters, ...
As steps to do failover/fallback are not covered, maybe explain that *automatic failover* will not be available (yet) either.
What other pieces are missing? Integration with other tools that sync objects between Kubernetes clusters?
This only discusses storage, pointers/links to how applications can be mirrored would be nice to have. No details, just enough to get an idea how a complete setup could be achieved at one point.

Author: > Are there any other requirements to the environment, like firewalls, response times between Ceph clusters, ...

Not planning to talk about mirror bootstrapping etc, maybe this can come as part of the usage document

> As steps to do failover/fallback are not covered, maybe explain that _automatic failover_ will not be available (yet) either.

step to do failover will be covered in usage document

> What other pieces are missing? Integration with other tools that sync objects between Kubernetes clusters?

Its kube admin responsibility to take application backup from one cluster and restore it on another cluster (from cephcsi point of view the PVC and PV objects should remain same)

> This only discusses storage, pointers/links to how applications can be mirrored would be nice to have. No details, just enough to get an idea how a complete setup could be achieved at one point.

We can add this to usage doc.

Ceph clusters. Mirroring ensures point-in-time, crash-consistent RBD images
between clusters, RBD mirroring is mainly used for disaster recovery (i.e.
having a secondary site as a failover). See [Ceph
documentation](https://docs.ceph.com/en/latest/rbd/rbd-mirroring) on RBD
mirroring for complete information.

## Architecture

![mirror](rbd-mirror.png)

## Design

Currently, CEPH-CSI generates its unique ID for each RBD image and stores the
corresponding PersistentVolume (PV) name and the unique ID mapping, It creates
the RBD image with the unique ID and returns the encoded value which contains
all the required information for other operations. For mirroring, the same RBD
image will be mirrored to the secondary cluster. As the journal(OMAP data) is
not mirrored to the secondary cluster, The RBD images corresponding to the PV
can not be identified without OMAP data.

**Pre-req** It's expected that the Kubernetes Admin/User will create the static
PersistentVolumeClaim (PVC) on the secondary site during the failover.

**Note:** when the static PVC created on the secondary site we cannot use the
VolumeHandle to identify the OMAP data nor the image anymore because as we have

> only PoolID and ClusterID in the VolumeHandle. we cannot identify the correct

Reviewer 2: *we* -> *We*

pool name from the PoolID because pool name will remain the same on both
clusters but not the PoolID even the ClusterID can be different on the
secondary cluster.

> > Sample PV spec which will be used by controller to regenerate OMAP data

Reviewer 2: *.. by controller * - which controller? If it is the **Replication Controller** in the diagram, reference to it by full name (everywhere).

If one of the components that will be developed for this design is this controller, it can be explained a little more as one of the deliverable components. Maybe it is good to list which components need to get developed from scratch, and which ones need some modifications/extensions.

Author: done added more details, we dont be designing the Replication Controller as of now, this will be a new controller which will run as a sidecar in the provisioner pod

```yaml

> csi:

Reviewer 2: indention looks weird

        pool: replicapool
        radosNamespace: ""
      volumeHandle: 0001-0009-rook-ceph-0000000000000002-0c23de1c-18fb-11eb-a903-0242ac110005
```

> **VolumeHandle** is the unique volume name returned by the CSI volume plugin’s
CreateVolume to refer to the volume on all subsequent calls.

Once the static PVC is created on the secondary cluster, the Kubernetes User
can try delete the PVC,expand the PVC or mount the PVC. Incase of mounting
(NodeStageVolume) we will get the volume context in RPC call but not in the
Delete/Expand Request. In Delete/Expand RPC request only the VolumeHandle
(`clusterID-poolID-volumeuniqueID`) will be sent where it contains the encoded
information related to ClusterID and PoolID. The VolumeHandle is not useful in
secondary cluster as the PoolID and ClusterID always may not be the same.

To solve this problem, We will have a new controller running as part of
provisioner pod which watchers for the PV objects. When a PV is created it will
extract the required information from the PV spec and it will regenerate the
OMAP data and also it will generate a new VolumeHandle
(`newclusterID-newpoolID-volumeuniqueID`) and it creates an OMAP object for

> mapping between old VolumeHandle and new VolumeHandle. whenever cephcsi gets a

Reviewer 2: *cephcsi* -> *Ceph-CSI* , please use the same spelling/capitalization everywhere
